game-core/src/test/java/games/strategy/triplea/attachments/TechAbilityAttachmentTest.java

```diff
@@ -0,0 +1,178 @@
+        () -> attachment.getIntInRange(name, "NaN", 10, false));
+    final Exception e2 = assertThrows(IllegalArgumentException.class,
+        () -> attachment.getIntInRange(name, "NaN", -10, true));
+    assertTrue(e1.getCause() instanceof NumberFormatException);
+    assertTrue(e2.getCause() instanceof NumberFormatException);
+    assertTrue(e1.getMessage().contains("NaN"));
+    assertTrue(e2.getMessage().contains("NaN"));
+  }
+
+  @Test
+  public void testGetIntInRange_tooHigh() {
+    final Exception e = assertThrows(GameParseException.class, () -> attachment.getIntInRange(name, "20", 10, false));
+    assertTrue(e.getMessage().contains("20"));
+    assertTrue(e.getMessage().contains("10"));
+    assertTrue(e.getMessage().contains(name));
+    assertTrue(e.getMessage().contains(customToString));
+  }
+
+  @Test
+  public void testGetIntInRange_negativeNoUndefined() {
+    final Exception e = assertThrows(GameParseException.class, () -> attachment.getIntInRange(name, "-1", 10, false));
+    assertTrue(e.getMessage().contains("-1"));
+    assertTrue(e.getMessage().contains("10"));
+    assertTrue(e.getMessage().contains(name));
+    assertTrue(e.getMessage().contains(customToString));
+  }
+
+  @Test
+  public void testGetIntInRange_negativeUndefined() {
+    final Exception e = assertThrows(GameParseException.class, () -> attachment.getIntInRange(name, "-2", 10, true));
+    assertTrue(e.getMessage().contains("-2"));
+    assertTrue(e.getMessage().contains("10"));
+    assertTrue(e.getMessage().contains(name));
+    assertTrue(e.getMessage().contains(customToString));
+  }
+
+  @Test
+  public void testGetIntInRange_validValues() throws Exception {
+    assertEquals(-1, attachment.getIntInRange(name, "-1", 10, true));
+    assertEquals(10, attachment.getIntInRange(name, "10", 10, true));
+    assertEquals(0, attachment.getIntInRange(name, "0", 10, false));
+    assertEquals(10, attachment.getIntInRange(name, "10", 10, false));
+  }
+
+  @Test
+  public void testGetUnitType() throws Exception {
+    assertEquals(dummyUnitType, attachment.getUnitType(testUnitType));
+    verify(list).getUnitType(testUnitType);
+    verify(list).getUnitType(any());
+  }
+
+
+  @Test
+  public void testGetUnitType_noValue() {
+    final String test = "someNonExistentKey";
+    final Exception e = assertThrows(GameParseException.class, () -> attachment.getUnitType(test));
+    verify(list).getUnitType(test);
+    verify(list).getUnitType(any());
+    assertTrue(e.getMessage().contains(test));
+  }
+
+  @Test
+  public void testApplyCheckedValue() throws Exception {
+    final Map<UnitType, Integer> map = new HashMap<>();
+    attachment.applyCheckedValue(name, "1:" + testUnitType, map::put);
+    assertEquals(1, map.size());
+    assertEquals(1, (int) map.get(dummyUnitType));
+  }
+
+  @Test
+  public void testSumIntegerMap() {
+    final IntegerMap<UnitType> map = spy(new IntegerMap<>());
+    when(map.getInt(dummyUnitType)).thenReturn(-1, 20, 300);
```

Comment:
This is a bit of a smell because we're having the same `IntegerMap` instance return three different values when looking up the same key.  I submitted an alternative approach in #3143.